API: missing hipfftPlan3d, hipfftDestroy, hipfftSetStream, hipfftGetVersion, hipfftExecD2Z, hipfftExecZ2D

The current `libhipfft.so` exports a partial set of symbols:

\`\`\`
T hipfftCreate
T hipfftExecC2C
T hipfftExecC2R
T hipfftExecR2C
T hipfftMakePlan1d
T hipfftMakePlan2d
T hipfftPlan1d
T hipfftPlan2d
\`\`\`

Missing symbols that AMD HIP's hipfft.h declares and that downstream applications need:

| Symbol | Used by |
|---|---|
| `hipfftPlan3d` | OpenMM PME (3D charge grid FFTs) |
| `hipfftMakePlan3d` | AMD HIP standard 3D plan creation |
| `hipfftDestroy` | All apps for cleanup of any plan |
| `hipfftSetStream` | Apps that run FFT on a dedicated stream |
| `hipfftGetVersion` | Version probing (used by OpenMM to gate hipFFT vs builtin fallback) |
| `hipfftExecD2Z` | Double-precision real→complex |
| `hipfftExecZ2D` | Double-precision complex→real |
| `hipfftExecZ2Z` | Double-precision complex↔complex (AMOEBA multipole) |

## Reproducer

Building OpenMM's HIP platform against this `libhipfft.so` fails at link time:

\`\`\`
ld: libOpenMMHIP.so: undefined reference to `hipfftSetStream'
ld: libOpenMMHIP.so: undefined reference to `hipfftExecD2Z'
ld: libOpenMMHIP.so: undefined reference to `hipfftDestroy'
ld: libOpenMMHIP.so: undefined reference to `hipfftExecZ2D'
ld: libOpenMMHIP.so: undefined reference to `hipfftPlan3d'
ld: libOpenMMHIP.so: undefined reference to `hipfftGetVersion'
\`\`\`

OpenMM uses the standard upstream pattern: `FIND_PACKAGE(HIPFFT CONFIG)` + `hipfftPlan3d` / `hipfftExecD2Z` / etc. — the same code that builds against AMD's hipFFT on ROCm.

--- a/src/hipfft.cpp
+++ b/src/hipfft.cpp
@@ -6,6 +6,7 @@
 #include <vector>
 #include "hipfftHandle.h"
 #include "hipfft.h"
+#include "hipfft-version.h"
 #include "hip/hip_runtime.h"
 #include "hip/hip_interop.h"
 #include "h4i/mklshim/mklshim.h"
@@ -263,3 +264,105 @@ hipfftResult hipfftExecC2C(hipfftHandle plan,
 
     return result;
 };
+
+
+hipfftResult hipfftPlan3d(hipfftHandle *plan, int nx, int ny, int nz, hipfftType type)
+{
+    if (nx <= 0 || ny <= 0 || nz <= 0) {
+        *plan = nullptr;
+        return HIPFFT_INVALID_VALUE;
+    }
+    hipfftHandle h = nullptr;
+    hipfftResult result = hipfftCreate(&h);
+    if (result != HIPFFT_SUCCESS) return result;
+    result = hipfftMakePlan3d(h, nx, ny, nz, type, nullptr);
+    *plan = h;
+    return result;
+}
+
+
+hipfftResult hipfftMakePlan3d(hipfftHandle plan, int nx, int ny, int nz, hipfftType type, size_t *workSize)
+{
+    std::vector<std::int64_t> dimensions {(int64_t)nx, (int64_t)ny, (int64_t)nz};
+    switch (type) {
+        case HIPFFT_R2C:
+        case HIPFFT_C2R:
+            plan->descSR = H4I::MKLShim::createFFTDescriptorSR(plan->ctxt, dimensions);
+            break;
+        case HIPFFT_D2Z:
+        case HIPFFT_Z2D:
+            plan->descDR = H4I::MKLShim::createFFTDescriptorDR(plan->ctxt, dimensions);
+            break;
+        case HIPFFT_C2C:
+            plan->descSC = H4I::MKLShim::createFFTDescriptorSC(plan->ctxt, dimensions);
+            break;
+        case HIPFFT_Z2Z:
+            plan->descDC = H4I::MKLShim::createFFTDescriptorDC(plan->ctxt, dimensions);
+            break;
+        default:
+            return HIPFFT_INVALID_PLAN;
+    }
+    return HIPFFT_SUCCESS;
+}
+
+
+hipfftResult hipfftExecZ2Z(hipfftHandle plan, hipfftDoubleComplex *idata, hipfftDoubleComplex *odata, int direction)
+{
+    switch (direction) {
+        case HIPFFT_FORWARD:
+            H4I::MKLShim::fftExecZ2Zforward(plan->ctxt, plan->descDC,
+                (double _Complex *)idata, (double _Complex *)odata);
+            return HIPFFT_SUCCESS;
+        case HIPFFT_BACKWARD:
+            H4I::MKLShim::fftExecZ2Zbackward(plan->ctxt, plan->descDC,
+                (double _Complex *)idata, (double _Complex *)odata);
+            return HIPFFT_SUCCESS;
+        default:
+            return HIPFFT_INVALID_VALUE;
+    }
+}
+
+
+hipfftResult hipfftExecD2Z(hipfftHandle plan, hipfftDoubleReal *idata, hipfftDoubleComplex *odata)
+{
+    H4I::MKLShim::fftExecD2Z(plan->ctxt, plan->descDR, idata, (double _Complex *)odata);
+    return HIPFFT_SUCCESS;
+}
+
+
+hipfftResult hipfftExecZ2D(hipfftHandle plan, hipfftDoubleComplex *idata, hipfftDoubleReal *odata)
+{
+    H4I::MKLShim::fftExecZ2D(plan->ctxt, plan->descDR, (double _Complex *)idata, odata);
+    return HIPFFT_SUCCESS;
+}
+
+
+hipfftResult hipfftDestroy(hipfftHandle plan)
+{
+    if (plan == nullptr) return HIPFFT_INVALID_PLAN;
+    if (plan->descSR != nullptr) H4I::MKLShim::destroyFFTDescriptorSR(plan->ctxt, plan->descSR);
+    if (plan->descSC != nullptr) H4I::MKLShim::destroyFFTDescriptorSC(plan->ctxt, plan->descSC);
+    if (plan->descDR != nullptr) H4I::MKLShim::destroyFFTDescriptorDR(plan->ctxt, plan->descDR);
+    if (plan->descDC != nullptr) H4I::MKLShim::destroyFFTDescriptorDC(plan->ctxt, plan->descDC);
+    delete plan;
+    return HIPFFT_SUCCESS;
+}
+
+
+// chipStar dispatches kernels on its own queue; the MKLShim context already
+// holds the SYCL queue used for the FFT.  hipFFT consumers (e.g. OpenMM) call
+// hipfftSetStream to associate a hipStream_t with the plan — accept it as a
+// no-op so existing code paths link and run; once MKLShim exposes a
+// stream-binding API this can be wired through.
+hipfftResult hipfftSetStream(hipfftHandle /*plan*/, hipStream_t /*stream*/)
+{
+    return HIPFFT_SUCCESS;
+}
+
+
+hipfftResult hipfftGetVersion(int *version)
+{
+    if (version == nullptr) return HIPFFT_INVALID_VALUE;
+    *version = hipfftVersionMajor * 10000 + hipfftVersionMinor * 100 + hipfftVersionPatch;
+    return HIPFFT_SUCCESS;
+}

--- a/tests/CMakeLists.txt
+++ b/tests/CMakeLists.txt
@@ -1,3 +1,4 @@
 add_subdirectory(hipfft_create)
 add_subdirectory(hipfft_real_1d)
 add_subdirectory(hipfft_complex_1d)
+add_subdirectory(hipfft_missing_symbols)

--- a/tests/hipfft_missing_symbols/CMakeLists.txt
+++ b/tests/hipfft_missing_symbols/CMakeLists.txt
@@ -1,0 +1,11 @@
+add_executable(hipfft_missing_symbols_test hipfft_missing_symbols.cpp)
+
+target_include_directories(hipfft_missing_symbols_test PRIVATE
+    ${CMAKE_SOURCE_DIR}/include
+)
+
+target_link_libraries(hipfft_missing_symbols_test PRIVATE hipfft)
+
+add_test(NAME hipfft_missing_symbols COMMAND hipfft_missing_symbols_test)
+set_tests_properties(hipfft_missing_symbols PROPERTIES
+    PASS_REGULAR_EXPRESSION "PASSED")

--- a/tests/hipfft_missing_symbols/hipfft_missing_symbols.cpp
+++ b/tests/hipfft_missing_symbols/hipfft_missing_symbols.cpp
@@ -1,0 +1,115 @@
+// Reproduces issue #8: missing hipfftPlan3d, hipfftMakePlan3d, hipfftDestroy,
+// hipfftSetStream, hipfftGetVersion, and the double-precision Exec variants
+// (Z2Z, D2Z, Z2D).  Today this fails to link because libhipfft.so doesn't
+// export these symbols.  After the fix it should run end-to-end.
+#include <cstdlib>
+#include <iostream>
+#include <vector>
+#include "hip/hip_runtime.h"
+#include "hipfftHandle.h"
+#include "hipfft.h"
+
+#define CHECK(call) do { \
+    hipfftResult r = (call); \
+    if (r != HIPFFT_SUCCESS) { \
+        std::cerr << "FAILED: " #call " returned " << r << std::endl; \
+        return EXIT_FAILURE; \
+    } \
+} while (0)
+
+int main() {
+    int version = 0;
+    CHECK(hipfftGetVersion(&version));
+    if (version <= 0) {
+        std::cerr << "FAILED: hipfftGetVersion returned non-positive " << version << std::endl;
+        return EXIT_FAILURE;
+    }
+
+    const int nx = 8, ny = 8, nz = 8;
+
+    // hipfftPlan3d: single-precision C2C
+    hipfftHandle plan3d_c2c = nullptr;
+    CHECK(hipfftPlan3d(&plan3d_c2c, nx, ny, nz, HIPFFT_C2C));
+
+    hipStream_t stream = nullptr;
+    if (hipStreamCreate(&stream) != hipSuccess) {
+        std::cerr << "FAILED: hipStreamCreate" << std::endl;
+        return EXIT_FAILURE;
+    }
+    CHECK(hipfftSetStream(plan3d_c2c, stream));
+
+    // 3D double-complex round trip via Z2Z
+    hipfftHandle plan3d_z2z = nullptr;
+    CHECK(hipfftPlan3d(&plan3d_z2z, nx, ny, nz, HIPFFT_Z2Z));
+
+    const size_t nelem = (size_t)nx * ny * nz;
+    std::vector<hipfftDoubleComplex> host_in(nelem), host_out(nelem);
+    for (size_t i = 0; i < nelem; i++) {
+        host_in[i].x = (double)(i % 7) - 3.0;
+        host_in[i].y = (double)(i % 5) - 2.0;
+    }
+    hipfftDoubleComplex *d_a = nullptr, *d_b = nullptr;
+    if (hipMalloc(&d_a, nelem * sizeof(hipfftDoubleComplex)) != hipSuccess ||
+        hipMalloc(&d_b, nelem * sizeof(hipfftDoubleComplex)) != hipSuccess) {
+        std::cerr << "FAILED: hipMalloc" << std::endl;
+        return EXIT_FAILURE;
+    }
+    hipMemcpy(d_a, host_in.data(), nelem * sizeof(hipfftDoubleComplex), hipMemcpyHostToDevice);
+    CHECK(hipfftExecZ2Z(plan3d_z2z, d_a, d_b, HIPFFT_FORWARD));
+    CHECK(hipfftExecZ2Z(plan3d_z2z, d_b, d_a, HIPFFT_BACKWARD));
+    hipMemcpy(host_out.data(), d_a, nelem * sizeof(hipfftDoubleComplex), hipMemcpyDeviceToHost);
+
+    // Round-trip with backend normalization: input ≈ output / N
+    double tol = 1e-6 * (double)nelem;
+    for (size_t i = 0; i < nelem; i++) {
+        double rx = host_out[i].x / (double)nelem;
+        double ry = host_out[i].y / (double)nelem;
+        if (std::abs(rx - host_in[i].x) > tol || std::abs(ry - host_in[i].y) > tol) {
+            std::cerr << "FAILED: Z2Z round-trip at i=" << i
+                      << " got (" << rx << "," << ry
+                      << ") expected (" << host_in[i].x << "," << host_in[i].y << ")" << std::endl;
+            return EXIT_FAILURE;
+        }
+    }
+
+    // 1D double real round trip via D2Z + Z2D.  Use 1D where the
+    // conjugate-even packed layout is unambiguous; the 3D layout
+    // (cuFFT full-complex vs oneMKL COMPLEX_REAL) is a separate
+    // storage-config concern out of scope for this PR.
+    const int n1d = 16;
+    hipfftHandle plan1d_d2z = nullptr, plan1d_z2d = nullptr;
+    CHECK(hipfftPlan1d(&plan1d_d2z, n1d, HIPFFT_D2Z, 1));
+    CHECK(hipfftPlan1d(&plan1d_z2d, n1d, HIPFFT_Z2D, 1));
+
+    double *d_r = nullptr;
+    hipfftDoubleComplex *d_c = nullptr;
+    hipMalloc(&d_r, (n1d + 2) * sizeof(double));
+    hipMalloc(&d_c, (n1d / 2 + 1) * sizeof(hipfftDoubleComplex));
+    std::vector<double> host_r_in(n1d), host_r_out(n1d);
+    for (int i = 0; i < n1d; i++) host_r_in[i] = (double)(i % 11) - 5.0;
+    hipMemcpy(d_r, host_r_in.data(), n1d * sizeof(double), hipMemcpyHostToDevice);
+    CHECK(hipfftExecD2Z(plan1d_d2z, d_r, d_c));
+    CHECK(hipfftExecZ2D(plan1d_z2d, d_c, d_r));
+    hipMemcpy(host_r_out.data(), d_r, n1d * sizeof(double), hipMemcpyDeviceToHost);
+    double tol1d = 1e-6 * (double)n1d;
+    for (int i = 0; i < n1d; i++) {
+        double r = host_r_out[i] / (double)n1d;
+        if (std::abs(r - host_r_in[i]) > tol1d) {
+            std::cerr << "FAILED: 1D D2Z/Z2D round-trip at i=" << i
+                      << " got " << r << " expected " << host_r_in[i] << std::endl;
+            return EXIT_FAILURE;
+        }
+    }
+
+    // hipfftDestroy should clean up all four plans without crashing
+    CHECK(hipfftDestroy(plan3d_c2c));
+    CHECK(hipfftDestroy(plan3d_z2z));
+    CHECK(hipfftDestroy(plan1d_d2z));
+    CHECK(hipfftDestroy(plan1d_z2d));
+
+    hipFree(d_a); hipFree(d_b); hipFree(d_r); hipFree(d_c);
+    hipStreamDestroy(stream);
+
+    std::cout << "PASSED" << std::endl;
+    return EXIT_SUCCESS;
+}

